hash_practice/exercises.py

```diff
@@ -1,29 +1,163 @@
+    times that letter appears in the string.
+    Time Complexity: O(n^2)
+    Space Complexity: O(n)
+    """
+    grouped_string_dicts = {}
+
+    for string in strings: 
+        string_dict = defaultdict(int)
+
+        for char in string:
+            string_dict[char] += 1
+
+        grouped_string_dicts[string] = string_dict
+
+    return grouped_string_dicts
+
 def grouped_anagrams(strings):
-    """ This method will return an array of arrays.
-        Each subarray will have strings which are anagrams of each other
-        Time Complexity: ?
-        Space Complexity: ?
+    """ 
+    Return an array of arrays.
+    Each subarray will have strings, which are anagrams of each other.
+    Time Complexity: O(n^2)
+    Space Complexity: O(n)
     """
-    pass
+    grouped_anagrams = []
+
+    if not strings:
+        return grouped_anagrams
+
+    grouped_string_dicts = create_string_freq_map(strings)
+
+    for string in strings:
+        anagram_array = []
+        anagram_array.append(string)
+
+        for word, char_freq_dict in grouped_string_dicts.items():
+            if grouped_string_dicts[string] == char_freq_dict and word != string:
+                anagram_array.append(word)
+                strings.remove(word)
+
+        grouped_anagrams.append(anagram_array)
+
+    return grouped_anagrams
 
 def top_k_frequent_elements(nums, k):
-    """ This method will return the k most common elements
-        In the case of a tie it will select the first occuring element.
-        Time Complexity: ?
-        Space Complexity: ?
+    """ 
+    Returns the k most common elements
+    In the case of a tie, it will select the first occuring element.
+    Time Complexity: O(n)
+    Space Complexity: O(n)
+    """
+    most_common_nums = []
+
+    if not nums:
+        return most_common_nums
+
+    freq_map = defaultdict(int)
+    for num in nums:
+        freq_map[num] += 1
+
+    while len(most_common_nums) < k:
+        most_freq_num = max(freq_map, key = lambda x: freq_map[x])
+        most_common_nums.append(most_freq_num)
+        del freq_map[most_freq_num]
+
+    return most_common_nums
+
+def valid_sudoku_rows(table):
+    """
+    Returns true if all rows in a sudoku table are valid.
+    Time Complexity: O(n^2)
+    Space Complexity: O(n)
+    """
+    for row in table:
+        row_set = set()
+
+        for char in row:
+            if char not in row_set and char != ".":
+                row_set.add(char)
+            elif char in row_set:
+                return False
+            
+    return True
+
+def valid_sudoku_columns(table):
     """
-    pass
+    Returns true if all columns in a sudoku table are valid.
+    Time Complexity: O(n)
+    Space Complexity: O(n)
+    """
+    index = 0
+
+    while index < 9: 
+        column_set = set()
+        
+        for row in table:
+            if row[index] not in column_set and row[index] != ".":
+                column_set.add(row[index])
+            

```

Comment:
Minor style: probably want to keep the `elif` next to its partner `if`. Just makes it more obvious they're the same "block".